### PR TITLE
Sage 300: Mapping (Base Branch) - 0

### PR DESCRIPTION
**Pull request by anishfyle**
## Summary by CodeRabbit

- **New Features**
  - Introduced detailed type definitions for various mapping models to enhance data structure clarity.
  - Added new components for Sage 300 mapping functionality, including routing and module configuration.
  - Implemented a new generic mapping table component with dynamic data display and user interaction enhancements.
  - Created a sub-menu component for improved navigation within the application.
  - Expanded the shared module with the addition of new components related to mapping and navigation.

- **Enhancements**
  - Improved the user interface for mapping functionalities with conditional rendering and new input/output bindings.
  - Enhanced pagination services with the ability to retrieve and store page sizes for different modules.
  - Streamlined the import paths for the `Paginator` model across various components.

- **Bug Fixes**
  - Corrected the property name from `destination_vendo` to `destination_vendor` in the employee mapping model.
  - Removed unnecessary whitespace and newline characters in the expense attribute model.

- **Refactor**
  - Reorganized the file structure for models, resulting in updated import paths for the `Paginator` model.
  - Simplified the usage of `this` in place of `that` within the `triggerAutoMapEmployees` method for better code clarity.

- **Style**
  - Applied Tailwind CSS utility classes for consistent styling across new and existing components.
  - Adjusted CSS styles for tab menu components to improve visual appeal and user experience.

- **Tests**
  - Added test suites for new components to ensure proper creation and initialization within the Angular application.

### CODE REVIEW

File: src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+  filteredMappings: ExtendedGenericMappingV2[];
+
+  searchTerm: string = '';
+
+  @Input() destinationOptions: DestinationAttribute[];
+
+  sourceType: string;
+
+  limit: number = 10;
```
Comment (ashwin1111) [inline src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts:50]: we shouldn't hard code limit, we should read it from localstorage

File: src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+  filteredMappingCount: number;
+
+  selectedMappingFilter: MappingState = MappingState.ALL;
+
+  PaginatorPage = PaginatorPage;
+
+  currentPage: number = 1;
+
+  searchValue: string;
```
Comment (ashwin1111) [inline src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts:64]: not used anywhere

File: src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+  destinationFieldType = FieldType;
```
Comment (ashwin1111) [inline src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts:66]: not used anywhere

File: src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+  operationgSystem: string;
```
Comment (ashwin1111) [inline src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts:68]: not used

File: src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+  alphabetFilter: string = 'All';
+
+  reimbursableExpenseObject?: IntacctReimbursableExpensesObject;
```
Comment (ashwin1111) [inline src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts:72]: not needed I think, remove it off

File: src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+  cccExpenseObject?: CorporateCreditCardExpensesObject;
```
Comment (ashwin1111) [inline src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts:74]: not needed I think, remove it off

File: src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+  constructor(
+    private mappingService: SiMappingsService,
```
Comment (ashwin1111) [inline src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts:77]: should be from common service

File: src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    private paginatorService: PaginatorService,
+    private route: ActivatedRoute,
+    private toastService: IntegrationsToastService
+  ) { }
+
+  triggerAutoMapEmployees() {
+    const that = this;
```
Comment (ashwin1111) [inline src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts:84]: we can emit it to it's parent again, that way the whole logic can live inside top level parent component

File: src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+      this.toastService.displayToastMessage(ToastSeverity.SUCCESS, 'Auto mapping of employees may take few minutes');
+    }, () => {
+      that.isLoading = false;
+      this.toastService.displayToastMessage(ToastSeverity.ERROR, 'Something went wrong, please try again');
+    });
+  }
+
+  private getFilteredMappings() {
+    this.mappingService.getGenericMappingsV2(this.limit, this.offset, this.destinationField, this.selectedMappingFilter, this.alphabetFilter, this.mappingPageName).subscribe((intacctMappingResult: GenericMappingV2Response) => {
+      this.filteredMappings = intacctMappingResult.results.concat();
```
Comment (ashwin1111) [inline src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts:97]: remove all intacct keywords

File: src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+  @Input() employeeFieldMapping: FyleField;
+
+  isInitialSetupComplete: boolean = false;
+
+  @Input() showAutoMapEmployee: boolean;
+
+  mappingStats: MappingStats;
+
+  employeeMapping: GenericMappingV2Response;
```
Comment (ashwin1111) [inline src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts:38]: not used I think

File: src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    this.isLoading = true;
+    this.selectedMappingFilter = state;
+    this.currentPage = 1;
+    this.offset = 0;
+    this.getFilteredMappings();
+  }
+
+  mappingSearchFilter(searchValue: string) {
+    if (searchValue.length > 0) {
+      const results: ExtendedGenericMappingV2[] = this.filteredMappings.filter((mapping) =>
```
Comment (ashwin1111) [inline src/app/shared/components/si/helper/generic-mapping-v2/generic-mapping-v2.component.ts:132]: this.filteredMappings =

File: src/app/core/models/si/db/generic-mapping-v2.model.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
+import { CategoryMapping } from './category-mapping.model';
+import { EmployeeMapping } from './employee-mapping.model';
+import { ExpenseAttributeDetail } from './expense-attribute-detail.model';
+import { MappingIntacct } from './mapping.model';
+
```
[the remark this answers is not in the document]
Comment (anishfyle) [reply src/app/core/models/si/db/generic-mapping-v2.model.ts:5]: duplicate, deleted [here](https://github.com/fylein/fyle-integrations-settings-app/pull/265)

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**